### PR TITLE
API: Search endpoint

### PR DESCRIPTION
**Pull request by annando**
The search endpoint is now enabled as well.

### CODE REVIEW

File: src/Module/Api/Mastodon/Accounts/Search.php [hunk 1]
```diff
@@ -44,13 +44,13 @@ public static function rawContent(array $parameters = [])
...
 
 		// What to search for
-		$q = (int)!isset($_REQUEST['q']) ? 0 : $_REQUEST['q'];
+		$q = $_REQUEST['q'] ?? '';
 		// Maximum number of results. Defaults to 40.
-		$limit = (int)!isset($_REQUEST['limit']) ? 40 : $_REQUEST['limit'];
+		$limit = (int)$_REQUEST['limit'] ?? 40;
 		// Attempt WebFinger lookup. Defaults to false. Use this when q is an exact address.
-		$resolve = (int)!isset($_REQUEST['resolve']) ? 0 : $_REQUEST['resolve'];
+		$resolve = (bool)!isset($_REQUEST['resolve']) ? false : ($_REQUEST['resolve'] == 'true');
```
Comment (MrPetovan) [inline src/Module/Api/Mastodon/Accounts/Search.php]:
```suggestion
		$resolve = ($_REQUEST['resolve'] ?? false) == 'true';
```

File: src/Module/Api/Mastodon/Accounts/Search.php [hunk 2]
```diff
@@ -44,13 +44,13 @@ public static function rawContent(array $parameters = [])
...
 		// Only who the user is following. Defaults to false.
-		$following = (int)!isset($_REQUEST['following']) ? 0 : $_REQUEST['following'];
+		$following = (bool)!isset($_REQUEST['following']) ? false : ($_REQUEST['following'] == 'true');
```
Comment (MrPetovan) [inline src/Module/Api/Mastodon/Accounts/Search.php]:
```suggestion
		$following = ($_REQUEST['following'] ?? false) == 'true';
```

File: src/Module/Api/Mastodon/Search.php [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+		// Return results immediately newer than this id
+		$min_id = (int)($_REQUEST['min_id'] ?? 0);
+		// Enum(accounts, hashtags, statuses)
+		$type = $_REQUEST['type'] ?? '';
+		// Filter out unreviewed tags? Defaults to false. Use true when trying to find trending tags.
+		$exclude_unreviewed = (bool)!isset($_REQUEST['exclude_unreviewed']) ? false : ($_REQUEST['exclude_unreviewed'] == 'true');
+		// The search query
+		$q = $_REQUEST['q'] ?? '';
+		// Attempt WebFinger lookup. Defaults to false.
+		$resolve = (bool)!isset($_REQUEST['resolve']) ? false : ($_REQUEST['resolve'] == 'true');
```
Comment (MrPetovan) [inline src/Module/Api/Mastodon/Search.php]:
```suggestion
		$resolve = ($_REQUEST['resolve'] ?? false) == 'true';
```

File: src/Module/Api/Mastodon/Search.php [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+		// Maximum number of results to load, per type. Defaults to 20. Max 40.
+		$limit = (int)($_REQUEST['limit'] ?? 20);
+		// Offset in search results. Used for pagination. Defaults to 0.
+		$offset = (int)($_REQUEST['offset'] ?? 0);
+		// Only who the user is following. Defaults to false.
+		$following = (bool)!isset($_REQUEST['following']) ? false : ($_REQUEST['following'] == 'true');
```
Comment (MrPetovan) [inline src/Module/Api/Mastodon/Search.php]:
```suggestion
		$following = ($_REQUEST['following'] ?? false) == 'true';
```

File: src/Module/Api/Mastodon/Search.php [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+		// If provided, statuses returned will be authored only by this account
+		$account_id = $_REQUEST['account_id'] ?? '';
+		// Return results older than this id
+		$max_id = (int)($_REQUEST['max_id'] ?? 0);
+		// Return results immediately newer than this id
+		$min_id = (int)($_REQUEST['min_id'] ?? 0);
+		// Enum(accounts, hashtags, statuses)
+		$type = $_REQUEST['type'] ?? '';
+		// Filter out unreviewed tags? Defaults to false. Use true when trying to find trending tags.
+		$exclude_unreviewed = (bool)!isset($_REQUEST['exclude_unreviewed']) ? false : ($_REQUEST['exclude_unreviewed'] == 'true');
```
Comment (MrPetovan) [inline src/Module/Api/Mastodon/Search.php]:
```suggestion
		$exclude_unreviewed = ($_REQUEST['exclude_unreviewed'] ?? false) == 'true';
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Comment by annando** [comment]
I added a new mechanism to check default values and presence of request parameters. This helps in two way:

1. a centralized check helps in applying changes if we have to adapt to some client behaviour (when for example a client is sending a boolean value as real boolean value)
2. We can easily detect if clients are transmitting undocumented parameters. I already experienced this. And I only found this by random.

I kept assigning the request array to variables. Doing so enables us to easily detect if we already handled that variable. Also we have a place to document the parameters.

I applied the changes only to two selected places since I'm awaiting your approval for that. Then I would convert all places.

**Review by MrPetovan — APPROVED** [review]

**Comment by annando** [comment]
I had expected some discussions :-) So you like this way?

**Comment by MrPetovan** [comment]
Huh, I didn't see what you did, I probably missed it in a partial diff, but it looks good to me.

**Pull request merged**